Assets/MixedRealityToolkit.SDK/Experimental/Features/Support/Scripts/BaseServiceManager.cs

```diff
@@ -0,0 +1,391 @@
+            for (int i = 0; i < providers.Length; i++)
+            {
+                // Check for null and mismatched type.
+                if ((providers[i] == null) || !interfaceType.IsAssignableFrom(providers[i].GetType())) { continue; }
+
+                // Check to see if name is valid (not null or whitespace) and if it matches the provider's name.
+                if (!string.IsNullOrWhiteSpace(name) && string.Equals(providers[i].Name, name))
+                {
+                    provider = (T)providers[i];
+                }
+                // If name is invalid, the first instance of a matching provider type indicates that a registration has occurred.
+                else
+                {
+                    provider = (T)providers[i];
+                }
+            }
+
+            return provider;
+        }
+
+        /// <inheritdoc />
+        public IReadOnlyList<T> GetDataProviders<T>(string name = null) where T : IMixedRealityDataProvider
+        {
+            if (!string.IsNullOrWhiteSpace(name))
+            {
+                Debug.LogError("This registrar does not support requesting multiple data providers of the same interface type and name.");
+                return new List<T>(); ;
+            }
+
+            Type interfaceType = typeof(T);
+            List<T> matchingProviders = new List<T>();
+
+            IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+            for (int i = 0; i < providers.Length; i++)
+            {
+                if (interfaceType.IsAssignableFrom(providers[i].GetType()))
+                {
+                    matchingProviders.Add((T)providers[i]);
+                }
+            }
+
+            return matchingProviders;
+        }
+
+        /// <inheritdoc />
+        public T GetService<T>(string name = null, bool showLogs = true) where T : IMixedRealityService
+        {
+            T serviceInstance = FindService<T>(name);
+
+            if (showLogs && (serviceInstance == null))
+            {
+                Debug.Log("Failed to get the requested service.");
+            }
+
+            return serviceInstance;
+        }
+
+        /// <inheritdoc />
+        public IReadOnlyList<T> GetServices<T>(string name = null) where T : IMixedRealityService
+        {
+            List<T> services = new List<T>();
+
+            if (!string.IsNullOrWhiteSpace(name))
+            {
+                Debug.LogError("This registrar does not support requesting multiple services of the same interface type and name.");
+                return services;
+            }
+
+            T serviceInstance = FindService<T>();
+            if (serviceInstance != null)
+            {
+                services.Add(serviceInstance);
+            }
+            return services;
+        }
+
+        /// <inheritdoc />
+        public bool IsDataProviderRegistered<T>(string name = null) where T : IMixedRealityDataProvider
```

Comment:
would be good to have documentation for all the public methods and properties

Reply:
These methods are documented as part of the IMixedRealityServiceRegistrar interface :)